### PR TITLE
Replace NetworkEndpoint with IstioEndpoint

### PR DESCRIPTION
**Pull request by nmittler**
The new IstioEndpoint was meant to replace
both ServiceInstance and NetworkEndpoint. This PR
removes the latter. ServiceInstance has been
refactored to use IstioEndpoint.

### CODE REVIEW

**Comment by nmittler** [comment]
@rshriram PTAL

**Comment by rshriram** [comment]
and fwiw 
>The new IstioEndpoint was meant to replace both ServiceInstance and NetworkEndpoint. 
was not the reason. the EDS code wanted its separate representation of the endpoint data. That is why :).

You are far better off moving IstioEndpoint into the network endpoint than the other way.

**Comment by nmittler** [comment]
@rshriram I didn't follow the entire history back ... I was simply reading the comment in the code which suggested that IstioEndpoint was meant to replace NetworkEndpoint. Do you know who originally introduced this? Probably worth a discussion with the author.

**Comment by rshriram** [comment]
Ignoring the who, the suggestion is to fold IstioEndpoint into NetworkEndpoint to avoid churning all these files and confining the changes to just proxy/envoy/v2/eds code. At the end of the day the functionality is the same.

**Comment by nmittler** [comment]
@rshriram 

> Ignoring the who, the suggestion is to fold IstioEndpoint into NetworkEndpoint to avoid churning all these files and confining the changes to just proxy/envoy/v2/eds code. At the end of the day the functionality is the same.

The [comment](https://github.com/istio/istio/blob/master/pilot/pkg/model/service.go#L322) literally says `TODO: Replace NetworkEndpoint and ServiceInstance with Istio endpoints`. 

Ultimately, the choice of `NetworkEndpoint` vs `IstioEndpoint` really doesn't matter much, however the fact that `IstioEndpoint` does not contain the service port is significant. When the kube controller fires endpoint events, it doesn't try to combine information from the service. If we went with NetworkEndpoint (as it is today), we would have to add that logic into the service controller, which may or may not have performance implications.  If we were to go the other way and remove the service port from `NetworkEndpoint`, it would effectively become `IstioEndpoint`. 

I believe this change is the correct one, both from a conceptual standpoint as well as being the original intent of `IstioEndpoint` as a replacement. If your concern is naming, I'd be happy to rename `IstioEndpoint` to something else (maybe just `Endpoint`).

File: pilot/pkg/config/coredatamodel/discovery_test.go [hunk 1]
```diff
@@ -142,51 +143,52 @@ func TestInstancesByPort(t *testing.T) {
...
 			servicePort: svcPort[0],
 			hostname:    host.Name("svc.example2.com"),
 		},
 	}
 
 	for _, svcInstance := range svcInstances {
 		step := steps[svcInstance.Endpoint.Address]
-		t.Run(fmt.Sprintf("verify service instances %s", step.address), func(_ *testing.T) {
+		t.Run(fmt.Sprintf("verify service instances %s", step.address), func(t *testing.T) {
+			g := gomega.NewGomegaWithT(t)
```
Comment (Nino-K) [inline pilot/pkg/config/coredatamodel/discovery_test.go]: Just curious why we are reinitializing gomega?

Comment (Nino-K) [reply pilot/pkg/config/coredatamodel/discovery_test.go]: I see it because of the new t.

File: pilot/pkg/model/service.go [hunk 2]
```diff
@@ -319,19 +267,19 @@ func GetLocalityOrDefault(label, defaultLocality string) string {
...
 // be available when endpoint callbacks are made.
-// - It no longer splits into one ServiceInstance and one NetworkEndpoint - both
-// are in a single struct
 // - doesn't have a pointer to Service - the full Service object may not be available at
 // the time the endpoint is received. The service name is used as a key and used to reconcile.
 // - it has a cached EnvoyEndpoint object - to avoid re-allocating it for each request and
 // client.
 type IstioEndpoint struct {
+	// Hostname of the service. Useful if the user needs to look up the Service from the endpoint.
+	Hostname host.Name
```
Comment (rshriram) [inline pilot/pkg/model/service.go]: ?? The PR is for renaming. If you want to tweak the model, let us do it in a separate PR.

Comment (howardjohn) [reply pilot/pkg/model/service.go]: It's not just renaming, right? We have all these structures that are subtly different for no real benefit. So to reconcile these it requires more than just renaming

**Comment by nmittler** [comment]
@rshriram PTAL. I'd really like to just get this in unless there is some serious objection.

**Comment by howardjohn** [comment]
I approved but put a hold to give @rshriram a chance to take a look

**Comment by nmittler** [comment]
@rshriram ping...

File: pilot/pkg/model/service.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/service.go]: Is this for future use?

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 3]
```diff
@@ -612,22 +612,27 @@ func (configgen *ConfigGeneratorImpl) buildInboundClusters(proxy *model.Proxy,
...
 				}
 			}
 
 			instance.Endpoint.Family = endpointFamily
 			instance.Endpoint.Address = endpointAddress
-			instance.Endpoint.ServicePort = listenPort
-			instance.Endpoint.Port = port
+			instance.ServicePort = listenPort
+			instance.Endpoint.ServicePortName = listenPort.Name
+			instance.Endpoint.EndpointPort = uint32(port)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: nit: no need uint32

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 4]
```diff
@@ -646,36 +651,30 @@ func (configgen *ConfigGeneratorImpl) buildInboundClusters(proxy *model.Proxy,
...
-		if realInstance.Endpoint.Port == int(ingressListener.Port.Number) {
-			instance = &model.ServiceInstance{
-				Endpoint:       realInstance.Endpoint,
-				Service:        realInstance.Service,
-				Labels:         realInstance.Labels,
-				ServiceAccount: realInstance.ServiceAccount,
-			}
-			break
+		if realInstance.Endpoint.EndpointPort == ingressListener.Port.Number {
+			instanceCopy := *realInstance
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Though this is not the business of this pr. But i think we should add a todo to evaluate if this shallow copy is meaningful and safe

File: pilot/pkg/model/service.go (hunk 2 above)
Comment (nmittler) [reply pilot/pkg/model/service.go]: It's not currently used, but it is logically needed in order to be afford callers the ability to map endpoints back to services. The only mechanism the API provides for searching for services is via hostname.  The controller always has the hostname available when constructing the endpoints, so it seems a low cost addition.

File: pilot/pkg/model/service.go [hunk 5]
```diff
@@ -206,56 +206,6 @@ const (
...
-//
-// For e.g., if catalog.mystore.com is accessible through port 80 and 8080,
-// and it maps to an instance with IP 172.16.0.1, such that connections to
-// port 80 are forwarded to port 55446, and connections to port 8080 are
-// forwarded to port 33333,
-//
-// then internally, we have two two endpoint structs for the
-// service catalog.mystore.com
-//  --> 172.16.0.1:54546 (with ServicePort pointing to 80) and
-//  --> 172.16.0.1:33333 (with ServicePort pointing to 8080)
```
Comment (rshriram) [inline pilot/pkg/model/service.go]: If you are going to delete, please move all these comments to  IstioEndpoint. This is useful info for new comers to understand.

File: pilot/pkg/model/service.go (hunk 2 above)
Comment (rshriram) [reply pilot/pkg/model/service.go]: This is not right. a pod can belong to multiple services. And this assumption is all over the place. So, please nuke this and we can investigate in a separate PR. The last thing I want is this piece being assumed a given by some clueless person writing the next main feature.

File: pilot/pkg/serviceregistry/external/conversion.go [hunk 6]
```diff
@@ -147,18 +147,24 @@ func convertEndpoint(service *model.Service, servicePort *networking.Port,
...
+			Family:          family,
+			EndpointPort:    instancePort,
+			ServicePortName: servicePort.Name,
+			Network:         endpoint.Network,
+			Locality:        endpoint.Locality,
+			LbWeight:        endpoint.Weight,
+			Labels:          endpoint.Labels,
+			TLSMode:         tlsMode,
+			Attributes: model.ServiceAttributes{
+				Name:      service.Attributes.Name,
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/external/conversion.go]: we should also nuke the name piece. Otherwise, we can't share the endpoint across different service objects. I also dont see where this is used.

Comment (nmittler) [reply pilot/pkg/serviceregistry/external/conversion.go]: That's beyond the scope of this PR. We'd have to give this more thought.

File: pilot/pkg/serviceregistry/external/servicediscovery.go [hunk 7]
```diff
@@ -93,23 +93,7 @@ func NewServiceDiscovery(configController model.ConfigStoreCache, store model.Is
 				instances := convertInstances(curr, cs)
 				endpoints := make([]*model.IstioEndpoint, 0)
 				for _, instance := range instances {
-					for _, port := range instance.Service.Ports {
-						endpoints = append(endpoints, &model.IstioEndpoint{
```
[the remark this answers is not in the document]
Comment (nmittler) [reply pilot/pkg/serviceregistry/external/servicediscovery.go]: @rshriram Are you questioning the changes or the original logic?

[the remark this answers is not in the document]
Comment (rshriram) [reply pilot/pkg/serviceregistry/external/servicediscovery.go]: You need an endpoint per service port per service instance. Your PR is simply adding an endpoint per instance for all ports

**Comment by howardjohn** [comment]
Fixes https://github.com/istio/istio/issues/10160
